bin/grep.py

```diff
@@ -3,22 +3,78 @@
 
 from __future__ import print_function
 
-import argparse
-import collections
-import fileinput
 import os
+import io
 import re
 import sys
+import argparse
+import fileinput
+import fnmatch
+from six import PY3
+
+if PY3:
+    from pathlib import Path
+else:
+    from pathlib2 import Path
+
+_stash = globals()['_stash']
+abbreviate = _stash.libcore.abbreviate
+
+new_file = True
+count = 0        
+ns = None
+fn = None
 
+def open_hook(filename,mode):
+    global fn
```

Comment:
Generally, you should avoid using globals. If you need an easy way to pass variables between functions, consider refactoring the code to use classes instead and change the globals into attributes.

Of course, global constants/functions and some other stuff is exempt from this, but it seems like you have a partial wrong understanding of the `global` statement. See my next comment for more info.

Reply:
Actually putting those global statements was more about documentation than functional necessity. I like to be explicit with my code and putting those global statements explicitly emphasised that those variables are meant to be the global ones and not some function-local variables. But you are right, global variables are generally not a good idea and those only made it into the code, because I started off with a bad example of using the file input module.